freqtrade/freqtradebot.py

```diff
@@ -704,7 +706,7 @@ def execute_entry(
         fee = self.exchange.get_fee(symbol=pair, taker_or_maker='maker')
         open_date = datetime.now(timezone.utc)
         if self.trading_mode == TradingMode.FUTURES:
-            funding_fees = self.exchange.get_funding_fees_from_exchange(pair, open_date)
+            funding_fees = self.exchange.get_funding_fees(pair, amount, open_date)
```

Comment:
Right now, we're only calling this twice - once "On entry" and once every hour(ish).

while i'm not convinced on calling this right after entry makes sense - i do think that we'll need to call this before/after exiting a trade.

it's pretty unlikely that the "update tick" happened in the 0.x seconds it takes us to go from "create_trade" to this line - and even if it did, it'd update latest within 15 minutes anyway.

On closing a trade on the other hand, we're not sure how the timing is ...so I think, we'll have to also call this right before/after closing a trade (maybe i'm wrong ... but it's a feeling i have whenever thinking about fees).

Reply:
thinking about this a little more, i think this is a pretty simple decision - we're either trusting schedule to work all the time (which means, the call right after entry is not necessary) - or we make sure we're calling it - in which case we should call it at entry - and right before trade exit.

Reply:
we need to call it when an order closes, is that inside the `_notify_enter` and `_notify_exit` methods?

Reply:
or is it in the `execute_entry` method, I already have a call there, I guess we only need to call it on exit if the scheduler doesn't do what it's supposed to for some reason

Reply:
I added in a safety update when the trade exits, and I tried to test it, but I'm getting an sql integrity error

```
elf = <sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x1316f1fa0>
cursor = <sqlite3.Cursor object at 0x131bbb500>
statement = 'INSERT INTO orders (ft_trade_id, ft_order_side, ft_pair, ft_is_open, order_id, status, symbol, order_type, side, pric...date, order_filled_date, order_update_date, leverage) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'
parameters = (1, 'sell', 'ETH/BTC', 1, 'mocked_limit_buy_usdt', 'open', ...)
context = <sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x131a069a0>

    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       sqlite3.IntegrityError: UNIQUE constraint failed: orders.ft_pair, orders.order_id

.env/lib/python3.9/site-packages/sqlalchemy/engine/default.py:717: IntegrityError

The above exception was the direct cause of the following exception:

mocker = <pytest_mock.plugin.MockerFixture object at 0x1328dfaf0>
default_conf = {'amend_last_stake_amount': False, 'ask_strategy': {'order_book_top': 1, 'price_side': 'ask', 'use_order_book': False}...delta': 1, 'enabled': False}, 'order_book_top': 1, 'price_side': 'bid', ...}, 'cancel_open_orders_on_exit': False, ...}
time_machine = <time_machine.TimeMachineFixture object at 0x13157b520>
fee = <MagicMock id='5122798064'>
ticker_usdt_sell_up = <MagicMock id='5127416176'>, is_short = False
limit_order_open = {'buy': {'amount': 30.0, 'cost': 60.0, 'datetime': '2021-11-10T07:20:52.498299+00:00', 'filled': 0.0, ...}, 'sell': {'amount': 30.0, 'cost': 66.0, 'datetime': '2021-11-10T07:20:52.498405+00:00', 'filled': 0.0, ...}}
schedule_off = True

    @pytest.mark.parametrize('schedule_off', [False, True])
    @pytest.mark.parametrize('is_short', [True, False])
    def test_update_funding_fees(
        mocker,
        default_conf,
        time_machine,
        fee,
        ticker_usdt_sell_up,
        is_short,
        limit_order_open,
        schedule_off
    ):
        '''
        nominal_value = mark_price * size
        funding_fee = nominal_value * funding_rate
        size = 123
        "LTC/BTC"
            time: 0, mark: 3.3, fundRate: 0.00032583, nominal_value: 405.9, fundFee: 0.132254397
            time: 8, mark: 3.2, fundRate: 0.00024472, nominal_value: 393.6, fundFee: 0.096321792
        "ETH/BTC"
            time: 0, mark: 2.4, fundRate: 0.0001, nominal_value: 295.2, fundFee: 0.02952
            time: 8, mark: 2.5, fundRate: 0.0001, nominal_value: 307.5, fundFee: 0.03075
        "ETC/BTC"
            time: 0, mark: 4.3, fundRate: 0.00031077, nominal_value: 528.9, fundFee: 0.164366253
            time: 8, mark: 4.1, fundRate: 0.00022655, nominal_value: 504.3, fundFee: 0.114249165
        "XRP/BTC"
            time: 0, mark: 1.2, fundRate: 0.00049426, nominal_value: 147.6, fundFee: 0.072952776
            time: 8, mark: 1.2, fundRate: 0.00032715, nominal_value: 147.6, fundFee: 0.04828734
        '''
        # SETUP
        time_machine.move_to("2021-09-01 00:00:00 +00:00")
    
        open_order = limit_order_open[enter_side(is_short)]
        bid = 0.11
        enter_rate_mock = MagicMock(return_value=bid)
        enter_mm = MagicMock(return_value=open_order)
        patch_RPCManager(mocker)
        patch_exchange(mocker)
        default_conf['trading_mode'] = 'futures'
        default_conf['collateral'] = 'isolated'
        default_conf['dry_run'] = True
    
        funding_rates = {
            "LTC/BTC": {
                1630454400000: 0.00032583,
                1630483200000: 0.00024472,
            },
            "ETH/BTC": {
                1630454400000: 0.0001,
                1630483200000: 0.0001,
            },
            "XRP/BTC": {
                1630454400000: 0.00049426,
                1630483200000: 0.00032715,
            }
        }
    
        mark_prices = {
            "LTC/BTC": {
                1630454400000: 3.3,
                1630483200000: 3.2,
            },
            "ETH/BTC": {
                1630454400000: 2.4,
                1630483200000: 2.5,
            },
            "XRP/BTC": {
                1630454400000: 1.2,
                1630483200000: 1.2,
            }
        }
    
        mocker.patch(
            'freqtrade.exchange.Exchange._get_mark_price_history',
            side_effect=lambda pair, since: mark_prices[pair]
        )
    
        mocker.patch(
            'freqtrade.exchange.Exchange.get_funding_rate_history',
            side_effect=lambda pair, since: funding_rates[pair]
        )
    
        mocker.patch.multiple(
            'freqtrade.exchange.Exchange',
            get_rate=enter_rate_mock,
            fetch_ticker=MagicMock(return_value={
                'bid': 1.9,
                'ask': 2.2,
                'last': 1.9
            }),
            create_order=enter_mm,
            get_min_pair_stake_amount=MagicMock(return_value=1),
            get_fee=fee,
        )
    
        freqtrade = get_patched_freqtradebot(mocker, default_conf)
    
        # initial funding fees,
        freqtrade.execute_entry('ETH/BTC', 123)
        freqtrade.execute_entry('LTC/BTC', 2.0)
        freqtrade.execute_entry('XRP/BTC', 123)
    
        trades = Trade.get_open_trades()
        assert len(trades) == 3
        for trade in trades:
            assert trade.funding_fees == (
                trade.amount *
                mark_prices[trade.pair][1630454400000] *
                funding_rates[trade.pair][1630454400000]
            )
    
        # create_mock_trades(fee, False)
        time_machine.move_to("2021-09-01 08:00:00 +00:00")
        if schedule_off:
            for trade in trades:
                assert trade.funding_fees == (
                    trade.amount *
                    mark_prices[trade.pair][1630454400000] *
                    funding_rates[trade.pair][1630454400000]
                )
>               freqtrade.execute_trade_exit(
                    trade=trade,
                    # The values of the next 2 params are irrelevant for this test
                    limit=ticker_usdt_sell_up()['bid'],
                    sell_reason=SellCheckTuple(sell_type=SellType.ROI)
                )

tests/test_freqtradebot.py:4838: 
```

Reply:
it's neither of the above - `notify_` methods are too late (and should not be used for this).

you should look at `execute_trade_exit` (or whatever that method is called now).

Reply:
> it's neither of the above - `notify_` methods are too late (and should not be used for this).
> 
> you should look at `execute_trade_exit` (or whatever that method is called now).
 
That's what i did

Reply:
@xmatthias I mean that that's what I did the first time, I haven't made any updates since your reply

Reply:
The reason itself is pretty simple - you're mocking `create_order()` and return a default object (which is fine for tests).
You're however never changing the mock to a "sell order" - which means that when the bot tries to sell the trade, it'll get the same order (the buy order) again - which fails to insert.

I think the simple fix for this is to recreate the `create_order` mock - but with a different order object (ideally one that's actually a sell - as i'm not sure if you'll not encounter other issues down that method ...).

Reply:
I've fixed that part now - but the test is still failing for different reasons (this falls into your realm i hope - i'm not quite sure why this fails).